Parallel planning example (#551)
ros-planning/moveit2_tutorials: doc/examples/parallel_planning/parallel_planning_tutorial.rst

> How to run multiple planning pipelines in parallel with MoveItCpp?

Reviewer: This isn't really a correct question
```suggestion
How to run multiple planning pipelines in parallel with MoveItCpp
```
A question would be:
> How do you run multiple planning pipelines in parallel with MoveItCpp?

Since these are titles, they don't need to be full sentences or questions. Removing the question mark should be fine.

Reviewer 2: I'd go even shorter: "Running Multiple Planning Pipelines in Parallel with MoveItCpp"

==================================================================

MoveItCpp offers an API that allows:

1. Running multiple planning pipelines in parallel with different planners
2. Defining a custom stopping criterion that terminates the pipelines that haven't found a solution
3. Defining a custom function to select the most suitable solutions

> Using multiple pipelines can be beneficial for example if:

Reviewer 2: ```suggestion
Using multiple pipelines can be beneficial for several reasons, including:
```

> - It is unsure which planner will produce the best solution for a given planning problem

Reviewer: "It" is a bit ambiguous here. On first read, I wasn't sure what "It" is referring to.

Reviewer 2: You could say something like "The planner that will produce the best solution is not known a priori"

- There is a chance that the preferred planner might fail and a backup solution should be available

> A general introduction to moveit_cpp can be found here: :doc:`/doc/examples/moveit_cpp/moveitcpp_tutorial`.

Reviewer: ```suggestion
A general introduction to moveit_cpp can be found in the :doc:`/doc/examples/moveit_cpp/moveitcpp_tutorial`.
```

Parallel Planning Interface
---------------------------

Using parallel planning with moveit_cpp is very similar to single pipeline planning except that a different implementation
of the planning component's :code:`plan(...)` function is used:

.. code-block:: c++

    MotionPlanResponse PlanningComponent::plan(
      const MultiPipelinePlanRequestParameters& parameters,
      SolutionCallbackFunction solution_selection_callback,
      StoppingCriterionFunction stopping_criterion_callback)

> This function tries to plan a trajectory from the a start state, to a goal state that satisfy a set of constraints. Based on the configuration

Reviewer 3: ```suggestion
This function tries to plan a trajectory from a start state, to a goal state that satisfy a set of constraints. Based on the configuration
```

> provided by the :code:`parameters` multiple threads are launched and each tries to solve the planning problem with a different planning pipeline. Once

Reviewer: ```suggestion
provided by the :code:`parameters`, multiple threads are launched and each tries to solve the planning problem with a different planning pipeline. Once
```

> all pipelines found a solution (Reminder: No solution is also a possible result), the :code:`solution_selection_callback` is called to determine which

Reviewer 3: ```suggestion
all pipelines terminate (Reminder: No solution is also a possible result), the :code:`solution_selection_callback` is called to determine which
```

> all pipelines found a solution (Reminder: No solution is also a possible result), the :code:`solution_selection_callback` is called to determine which

Reviewer: ```suggestion
all pipelines have found a solution (Reminder: No solution is also a possible result), the :code:`solution_selection_callback` is called to determine which
```

> solution is returned as :code:`MotionPlanResponse`. By default all pipelines use their time budget defined by the :code:`planning_time` field of the :code:`PlanRequestParameters` but it is possible to terminate the parallel planning earlier by using the :code:`stopping_criterion_callback`. This function

Reviewer: ```suggestion
solution is returned as :code:`MotionPlanResponse`. By default, all pipelines use their time budget defined by the :code:`planning_time` field of the :code:`PlanRequestParameters`, but it is possible to terminate the parallel planning earlier by using the :code:`stopping_criterion_callback`. This function
```

> is iteratively called during the parallel planning process and terminates pipelines that have not found a solution yet, if the stopping criterion is met.

Reviewer: ```suggestion
is iteratively called during the parallel planning process and, if the stopping criterion is met, terminates pipelines that have not found a solution yet.
```

Demo
----

> The following demo serves as an example how you can configure and use MoveItCpp's parallel planning interface. First of all, let's

Reviewer: ```suggestion
The following demo serves as an example of how you can configure and use MoveItCpp's parallel planning interface. First of all, let's
```

run the demo: ::

  ros2 launch moveit2_tutorials parallel_planning_example.launch.py

> A complex kitchen scene is loaded and two planning problems solved. The first one is a small motion of the EE towards the ground. This problem is likely to be solved by all three

Reviewer 3: ```suggestion
A complex kitchen scene is loaded and two planning problems solved. The first one is a small motion of the end effector towards the ground. This problem is likely to be solved by all three
```

> A complex kitchen scene is loaded and two planning problems solved. The first one is a small motion of the EE towards the ground. This problem is likely to be solved by all three

Reviewer: ```suggestion
A complex kitchen scene is loaded and two planning problems are solved. The first one is a small motion of the EE towards the ground. This problem is likely to be solved by all three
```

planners but with significant differences in the planning time. The second problem is a much harder and most likely only the :code:`RRTConnect` planner will succeed. This demo
suggests that a well configured parallel planning setup is versatile to be used in a broad variety of motion planning problems.

> What code is necessary to use parallel planning?

Reviewer: What do you think about making this a new subsection?

.. code-block:: c++

    // Stop parallel planning as soon as RRTConnect finds a solution
    bool stoppingCriterion(
        moveit_cpp::PlanningComponent::PlanSolutions const& plan_solutions,
        moveit_cpp::PlanningComponent::MultiPipelinePlanRequestParameters const& plan_request_parameters)
    {
      // Read solutions that are found up to this point from a thread safe storage
      auto const& solutions = plan_solutions.getSolutions();

      // Stop parallel planning if the pipeline using RRTConnect finds a solution
      for (auto const& solution : solutions)
      {
          if (solution.planner_id_ == "RRTConnectkConfigDefault")
          {
            // Return true to abort the other pipelines
            return true;
          }
      }
      // Return false when parallel planning should continue
      return false;
    }

Once :code:`MultiPipelinePlanRequestParameters` and optionally :code:`SolutionCallbackFunction` and/or :code:`StoppingCriterionFunction` are defined, we call :code:`plan(...)`:

.. code-block:: c++

    auto plan_solution = planning_component_->plan(multi_pipeline_plan_request, &getShortestSolution);

Tips
----

> - When you want to use multiple pipelines with the same planner parallel it is recommended to initialize multiple planning pipelines in moveit_cpp rather than using the same one in multiple parallel planning requests

Reviewer 3: Is this what you are saying here?:

>If you want to use multiple planners from the same planning pipeline (e.g., PRM and RRT from the OMPL planning pipeline), then it is recommended to use multiple instances of that planning pipeline.

Why "recommended"? What will break if you don't? Why is it not required?

Author: Yes, I rephrased the sentence to:
```
When you want to use different planners of the same pipeline (e.g. PILZ planner with PTP and LIN) in parallel, it is recommended to initialize multiple planning pipelines in moveit_cpp rather than using the same one in multiple parallel planning requests
```
I'd rather say recommend than required since it can be possible that a planner plugin can serve multiple requests at the same time (it seems to work with OMPL) but depending on the planner it could also result in undefined behavior, race conditions or "hidden sequential planning". Using two different pipelines in this case is the safest option for now but it would be nice to investigate the parallelism/concurrency capabilities of our plugins more in the future

Author: I've added a second OMPL pipeline that uses RRT*
